huaweicloud/services/vod/resource_huaweicloud_vod_watermark_template.go

```diff
@@ -0,0 +1,336 @@
+	var imageProcessResult vod.CreateWatermarkTemplateReqImageProcess
+	switch imageProcess {
+	case "ORIGINAL":
+		imageProcessResult = vod.GetCreateWatermarkTemplateReqImageProcessEnum().ORIGINAL
+	case "TRANSPARENT":
+		imageProcessResult = vod.GetCreateWatermarkTemplateReqImageProcessEnum().TRANSPARENT
+	case "GRAYED":
+		imageProcessResult = vod.GetCreateWatermarkTemplateReqImageProcessEnum().GRAYED
+	default:
+		return nil
+	}
+	return &imageProcessResult
+}
+
+func buildCreatePositionOpts(position string) *vod.CreateWatermarkTemplateReqPosition {
+	var positionResult vod.CreateWatermarkTemplateReqPosition
+	switch position {
+	case "TOPRIGHT":
+		positionResult = vod.GetCreateWatermarkTemplateReqPositionEnum().TOPRIGHT
+	case "TOPLEFT":
+		positionResult = vod.GetCreateWatermarkTemplateReqPositionEnum().TOPLEFT
+	case "BOTTOMLEFT":
+		positionResult = vod.GetCreateWatermarkTemplateReqPositionEnum().BOTTOMLEFT
+	case "BOTTOMRIGHT":
+		positionResult = vod.GetCreateWatermarkTemplateReqPositionEnum().BOTTOMRIGHT
+	default:
+		return nil
+	}
+	return &positionResult
+}
+
+func uploadImage(uploadUrl, fileName string) error {
+	data, err := os.Open(fileName)
+	if err != nil {
+		return err
+	}
+	defer data.Close()
+	req, err := http.NewRequest("PUT", uploadUrl, data)
+	if err != nil {
+		return err
+	}
+	req.Header.Set("Content-Type", "text/plain")
+
+	httpClient := &http.Client{}
+	res, err := httpClient.Do(req)
+	if err != nil {
+		return err
+	}
+	defer res.Body.Close()
+
+	return nil
+}
+
+func resourceWatermarkTemplateCreate(ctx context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
+	config := meta.(*config.Config)
+	client, err := config.HcVodV1Client(config.GetRegion(d))
+	if err != nil {
+		return diag.Errorf("error creating VOD client: %s", err)
+	}
+
+	createOpts := vod.CreateWatermarkTemplateReq{
+		Name:             d.Get("name").(string),
+		Type:             d.Get("image_type").(string),
+		ImageProcess:     buildCreateImageProcessOpts(d.Get("image_process").(string)),
+		Dx:               utils.String(d.Get("horizontal_offset").(string)),
+		Dy:               utils.String(d.Get("vertical_offset").(string)),
+		Position:         buildCreatePositionOpts(d.Get("position").(string)),
+		Width:            utils.String(d.Get("width").(string)),
+		Height:           utils.String(d.Get("height").(string)),
+		TimelineStart:    utils.String(d.Get("timeline_start").(string)),
+		TimelineDuration: utils.String(d.Get("timeline_duration").(string)),
+	}
+	log.Printf("[DEBUG] Create VOD watermark template Options: %#v", createOpts)
+
+	createReq := vod.CreateWatermarkTemplateRequest{
+		Body: &createOpts,
+	}
+
+	resp, err := client.CreateWatermarkTemplate(&createReq)
+	if err != nil {
+		return diag.Errorf("error creating VOD watermark template: %s", err)
+	}
+
+	err = uploadImage(*resp.UploadUrl, d.Get("image_file").(string))
+	if err != nil {
+		return diag.Errorf("error uploading watermark image: %s", err)
```

Comment:
should we move d.SetId above this uploadImage operation? if uploadImage or confirmImageUpload request errored, the resource creteated will be a dangling resource.

Reply:
sure